virtualmachine/gcp/util.go

```diff
@@ -435,3 +442,284 @@ func (svc *googleService) insertSSHKey() error {
+		if attachedDisk.Source == gDisk.SelfLink {
+			deviceName = attachedDisk.DeviceName
+			break
+		}
+	}
+
+	op, err := svc.service.Instances.DetachDisk(svc.vm.Project, svc.vm.Zone,
+		svc.vm.Name, deviceName).Do()
+
+	if err != nil {
+		return err
+	}
+	return svc.waitForOperationReady(op.Name)
+
+}
+
+// getFirewall gets details of a firewall
+func (svc *googleService) getFirewall() (*googlecloud.Firewall, error) {
+	return svc.service.Firewalls.Get(svc.vm.Project, svc.vm.Firewall).Do()
+}
+
+// addFirewallPorts adds new ports to a firewall
+func (svc *googleService) addFirewallPorts() error {
+
+	currFirewall, err := svc.getFirewall()
+	if err != nil {
+		return err
+	}
+	newRules := currFirewall.Allowed
+	for _, endpoint := range svc.vm.Endpoints {
+		newRules = append(newRules, &googlecloud.FirewallAllowed{
+			IPProtocol: endpoint.Protocol,
+			Ports:      endpoint.Ports,
+		})
+	}
+
+	return svc.patchFirewall(newRules)
+}
+
+// patchFirewall patches the firewall with the given allowed ports (rules)
+func (svc *googleService) patchFirewall(allowed []*googlecloud.FirewallAllowed) error {
+	newFirewall := &googlecloud.Firewall{
+		Allowed: allowed,
+	}
+	op, err := svc.service.Firewalls.Patch(svc.vm.Project, svc.vm.Firewall,
+		newFirewall).Do()
+	if err != nil {
+		return err
+	}
+	return svc.waitForGlobalOperationReady(op.Name)
+}
+
+// removeFirewallPorts removes ports from a given firewall
+func (svc *googleService) removeFirewallPorts() error {
+	firewall, err := svc.getFirewall()
+	if err != nil {
+		return err
+	}
+	// Initialize new rules with the current one
+	// Refer to this link for structure of firewall resource
+	// https://cloud.google.com/compute/docs/reference/latest/firewalls
+	newRules := firewall.Allowed
+	// Define a map to identify ports to be removed
+	remPorts := make(map[string]map[string]bool)
+	// Define a map to deal with elements where the same protocol
+	// appears again
+	encountered := make(map[string]bool)
+	// Identify ports for each protocol to be removed
+	for _, endpoint := range svc.vm.Endpoints {
+		// Initialize the map only if the given protocol is encountered
+		// for the first time
+		if !encountered[endpoint.Protocol] {
+			remPorts[endpoint.Protocol] = make(map[string]bool)
+			encountered[endpoint.Protocol] = true
+		}
+		for _, port := range endpoint.Ports {
+			remPorts[endpoint.Protocol][port] = true
+		}
+	}
+
+	// Remove the the ports and protocols that are to be removed from the
+	// rules. Iterate through each item in the list of rules.
+	for indEndp := 0; indEndp < len(newRules); indEndp++ {
+		allowed := newRules[indEndp]
+		// Iterate through each port in the port list
+		for indPort := 0; indPort < len(newRules[indEndp].Ports); indPort++ {
+			port := newRules[indEndp].Ports[indPort]
+			// Process removal if the given port is to be removed
+			if remPorts[allowed.IPProtocol][port] {
+				newRules[indEndp].Ports[indPort] = ""
+				newRules[indEndp].Ports = append(
```

Comment:
why do we need both this & above statement to remove the port entry?

Reply:
I read at https://github.com/golang/go/wiki/SliceTricks that "some elements with values are still referenced by slice and thus can not be collected". Thus for garbage collection I thought it is better to set the element to it's zero value. Let me know if my understanding is not correct and if this needs to be modified.